qiskit/optimization/algorithms/recursive_minimum_eigen_optimizer.py

```diff
@@ -14,23 +14,62 @@
 
 """A recursive minimal eigen optimizer in Qiskit's optimization module."""
 
-import logging
 from copy import deepcopy
-from typing import Optional
-
+from enum import Enum
+from typing import Optional, Union, List, Tuple, Dict
+import logging
 import numpy as np
+
 from qiskit.aqua.algorithms import NumPyMinimumEigensolver
 from qiskit.aqua.utils.validation import validate_min
 
-from .minimum_eigen_optimizer import MinimumEigenOptimizer, MinimumEigenOptimizerResult
 from .optimization_algorithm import OptimizationAlgorithm, OptimizationResult
+from .minimum_eigen_optimizer import MinimumEigenOptimizer, MinimumEigenOptimizationResult
 from ..converters.quadratic_program_to_qubo import QuadraticProgramToQubo
 from ..exceptions import QiskitOptimizationError
+from ..problems import Variable
 from ..problems.quadratic_program import QuadraticProgram
 
 logger = logging.getLogger(__name__)
 
 
+class IntermediateResult(Enum):
+    """
+    Defines whether the intermediate results at each iteration should be stored
+    and returned to the end user.
+    """
+
+    NO_ITERATIONS = 0
+    """No intermediate results are stored."""
+
+    LAST_ITERATION = 1
+    """Only results from the last iteration are stored."""
+
+    ALL_ITERATIONS = 2
+    """All intermediate results are stored."""
+
+
+class RecursiveMinimumEigenOptimizationResult(OptimizationResult):
+    """Recursive Eigen Optimizer Result."""
+    def __init__(self, x: Union[List[float], np.ndarray], fval: float,
+                 variables: List[Variable],
+                 replacements: Dict[str, Tuple[str, int]],
+                 history: List[OptimizationResult]) -> None:
+        super().__init__(x, fval, variables, None)
+        self._replacements = replacements
+        self._history = history
+
+    @property
+    def replacements(self) -> Dict[str, Tuple[str, int]]:
+        """Returns replacements."""
```

Comment:
Can we have a description of what this dictionary holds/is so the user has an idea of it contents?

Reply:
Added a short description.